### ISSUE TITLE
Error: reading Subscription Alias "subName": subscription.AliasClient#Get: Failure responding to request: StatusCode=401

### ISSUE THREAD

**Issue by danabrash** [hashicorp/terraform-provider-azurerm#12513]
Hello!

Terraform v1.0.1
hashicorp/azurerm v2.65.0 

I have a problem where my code to manage subscriptions works fine on my local machine, but it breaks when I run it in an ADO pipeline.

We're using a service principal to execute, and the service principal works fine on other pipelines executing TF.

This is the failing code block:

>   resource "azurerm_subscription" "subscriptions" {
>     for_each              = { for key, val in local.subscriptions : 
>                                 key => val if (
>                                     contains (
>                                       local.enabled_subscriptions,
>                                       val.subscription_name
>                                       ) == false
>                                     )
>                             }
>     subscription_name     = each.value.subscription_name
>     workload              = each.value.workload
>     billing_scope_id      = data.azurerm_billing_mca_account_scope.\<billingscope\>.id
>   }
> 

error in pipeline

> Error: reading Subscription Alias "\<subID\>": subscription.AliasClient#Get: Failure responding to request: StatusCode=401 -- Original Error: autorest/azure: Service returned an error. Status=401 Code="UserNotAuthorized" Message="User does not have access Microsoft.Subscription/aliases/read over scope providers/Microsoft.Subscription/aliases/<subID\>"
> 
>   with azurerm_subscription.subscriptions["\<subname\>"],
>   on main.tf line 17, in resource "azurerm_subscription" "subscriptions":
>   17: resource "azurerm_subscription" "subscriptions" {
> 

even:

> resource "azurerm_subscription" "\<subname\>" {
>   alias             = "\<subname\>"
>   subscription_name = "\<subname\>"
>   subscription_id   = "\<subID\>"
> }

errors with the same:

> Error: reading Subscription Alias "subName": subscription.AliasClient#Get: Failure responding to request: StatusCode=401 -- Original Error: autorest/azure: Service returned an error. Status=401 Code="UserNotAuthorized" Message="User does not have access Microsoft.Subscription/aliases/read over scope providers/Microsoft.Subscription/aliases/subName"
> 
>   with azurerm_subscription.subName,
>   on providers.tf line 61, in resource "azurerm_subscription" "subName":
>   61: resource "azurerm_subscription" "subName" {
> 

Expected result: subscriptions should be managed successfully as they were in development and testing

We've given the service principal owner on the tenant root mgmt group and made sure that cascaded down to the target subscriptions

I'm not sure where the failure is when my service principal is owner.

Thanks for any help.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by calexandre** [comment]
Hello, I'm also being affected by the same issue...

I don't know if it helps, but you can check the subscription aliases you have access to by running this command:

```powershell
(Invoke-AzRestMethod -Method "GET" -Path "/providers/Microsoft.Subscription/aliases?api-version=2020-09-01").Content | convertfrom-json | ConvertTo-Json -Depth 10
```

In my case, the subscription that returns `401` is not showing in the result, although I'm the owner of it...

**Comment by BrewToR** [comment]
Hello, I have the same issue. Did you fixed it ?
thanks !

**Comment by dnozay** [comment]
My coworker pointed out to me: https://docs.microsoft.com/en-us/azure/role-based-access-control/elevate-access-global-admin

TLDR:
> As a Global Administrator in Azure Active Directory (Azure AD), you might not have access to all subscriptions and management groups in your directory. This article describes the ways that you can elevate your access to all subscriptions and management groups.

It solved my case which was this type of issue:

```
Error: reading Subscription Alias "subName": subscription.AliasClient#Get: Failure responding to request: 
StatusCode=401 -- Original Error: autorest/azure: Service returned an error. Status=401 Code="UserNotAuthorized" 
Message="User does not have access Microsoft.Subscription/aliases/read over scope 
providers/Microsoft.Subscription/aliases/some-alias-name"
```

**Comment by tikicoder** [comment]
Does anyone know why the azurerm_subscription uses a subscription alias and not just the subscription ID?

**Comment by tombuildsstuff** [comment]
@tikicoder because the most reliable way to create a Subscription in ARM is to use a Subscription Alias - therefore you're managing the Subscription Alias (which creates a Subscription) rather than a Subscription directly.

**Comment by tombuildsstuff** [comment]
👋 

Taking a look through here since this is a permissions issue (as a 401 Unauthorized is being returned from the API) - I'm going to close this issue for the moment but [I believe this comment by @dnozay should help add these missing permissions](https://github.com/hashicorp/terraform-provider-azurerm/issues/12513#issuecomment-1014957043).

Thanks!

**Comment by tikicoder** [comment]
You can leave this closed, however, I disagree with the response and action.
Accepting the response of just be a Global Admin doesn't really fix the issue. From a security standpoint, how is assigning everyone global admin roll acceptable in the least trust practice? I called MS out on that as well as they gave me the same response.

How is the most reliable way to create a subscription in ARM is to use a subscription alias, my understanding from MS is this is a newer feature. I mean behind the curtain it creates a subscription and then added the alias. The part of being created with the Alias isn't the issue, its the part of it using it for reference.

(Looks like it was added in 2019) 
https://docs.microsoft.com/en-us/azure/templates/microsoft.subscription/2019-10-01-preview/aliases?tabs=bicep

The first non preview api was 2020.
https://docs.microsoft.com/en-us/azure/templates/microsoft.subscription/2020-09-01/aliases?tabs=bicep

Even talking with MS reps they were unsure about why aliases. There is no good way to manage permissions around who has access to Subscription aliases, the only way I have seen is either Global Admin or Owner. If you are not one then you do not have access.

Here is the issue. If you are working in an environment with multiple users and they are not  Global Administrators you will have issues. If you are using a pipeline with a service account where everything goes through the service account you would be fine.

Subscription aliases seem to live outside of the subscription so you as the creator/owner have access but unless you are a global admin no one else does. I even tried to create a custom role assignment for alias/* to get at least read permission, and it failed.
